src/mpid/ch4/src/ch4_progress.h

```diff
@@ -13,13 +13,100 @@
 
 #include "ch4_impl.h"
 
+#undef FUNCNAME
+#define FUNCNAME MPIDI_Progress_hooks
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_Progress_hooks(void)
+{
+    int mpi_errno, made_progress, num_of_hooks, i;
+    mpi_errno = MPI_SUCCESS;
+
+    num_of_hooks = OPA_load_int(&MPIDI_CH4_Global.registered_progress_hooks);
+    for (i = 0; i < num_of_hooks; i++) {
+        progress_func_ptr_t func_ptr = NULL;
+        MPID_THREAD_CS_ENTER(POBJ, MPIDI_CH4I_THREAD_PROGRESS_HOOK_MUTEX);
+        MPID_THREAD_CS_ENTER(VCI, MPIDI_CH4I_THREAD_PROGRESS_HOOK_MUTEX);
```

Comment:
This locking model is very weird.  First, why do you need both `POBJ` and `VCI` locks?  Second, the hooks themselves do not take a `vci` argument, which means that they are not vci-specific.  So, shouldn't you be getting some sort of a global lock, instead of either `POBJ` or `VCI` locks?

Reply:
(1) There's both the `POBJ` and `VCI` locking macros to handle both the `per-object` and `per-vci` critical sections. Right now, when `per-vci` is selected at configure time, the `POBJ` just expands to nothing (https://github.com/pmodels/mpich/blob/master/src/mpid/common/thread/mpidu_thread_fallback.h#L152). I think the reason why `MPID_THREAD_CS_ENTER_VCI*` locking macros were introduced in addition to the `MPID_THREAD_CS_ENTER_POBJ*` macros is that a working `per-object` critical section haven't been maintained for years.
(2) Yes, the hooks are not vci-specific. The lock being taken now is not a VCI-specific lock either. I guess the lock to be taken here could be the global lock but I am not sure if there is a reason why `MPIDI_CH4I_THREAD_PROGRESS_HOOK_MUTEX` is taken over the `MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX`.

Seems like the git diff is not doing a good job of showing what changes I introduced. This locking model was there before my changes.

Reply:
I think the first step would be delete old unmaintained code (e.g., `POBJ` locks).  Second, I think the `VCI` lock should be changed to `GLOBAL` lock for those hooks.